signac/core/file_buffered_synced_collection.py

```diff
@@ -0,0 +1,283 @@
+# Copyright (c) 2020 The Regents of the University of Michigan
+# All rights reserved.
+# This software is licensed under the BSD 3-Clause License.
+"""Implement the FileBufferedSyncedCollection class.
+
+The FileBufferedSyncedCollection is a concrete implementation of the buffer
+protocol established by the BufferedSyncedCollection class. It uses an
+in-memory cache to store data when in buffered mode. It is suitable for
+use with any file-based backend because it performs integrity checks based on
+whether or not the underlying file has been modified while buffering was
+activated.
+"""
+
+import errno
+import hashlib
+import json
+import os
+import sys
+
+from typing import Any, Dict
+
+from .buffered_synced_collection import BufferedCollection
+from .errors import MetadataError
+
+
+def get_buffer_capacity():
+    """Get the current buffer capacity."""
+    return FileBufferedCollection._BUFFER_CAPACITY
+
+
+def set_buffer_capacity(new_capacity):
+    """Update the buffer capacity."""
+    FileBufferedCollection._BUFFER_CAPACITY = new_capacity
+
+
+def get_current_buffer_size():
+    """Get the total amount of data currently stored in the buffer."""
+    return FileBufferedCollection._CURRENT_BUFFER_SIZE
+
+
+class FileBufferedCollection(BufferedCollection):
+    """Implement buffering for SyncedCollections with file-based backends.
+
+    All file-based backends can use the same set of integrity checks prior to a
+    buffer flush. This class standardizes that protocol.
+    """
+
+    _cache: Dict[str, Any] = {}
```

Comment:
See [conversation about bytes](https://github.com/glotzerlab/signac/pull/363#discussion_r545540825).
```suggestion
    _cache: Dict[str, bytes] = {}
```

Reply:
This isn't quite right, but I can try to make this more precise. The cache contains more than just the blob, it also contains the hash and the metadata, so if we really want to be precise I think it's `Dict[str, Dict[str, Union[bytes, str, Tuple[int, int]]]]`.